src/Security/ZitadelAuthenticator.php

```diff
@@ -0,0 +1,163 @@
+            'code' => $request->query->get('code'),
+        ]);
+
+        $rememberBadge = new RememberMeBadge();
+        $rememberBadge = $rememberBadge->enable();
+
+        return new SelfValidatingPassport(
+            new UserBadge($accessToken->getToken(), function () use ($accessToken, $client, $slugger) {
+                /** @var ZitadelResourceOwner $zitadelUser */
+                $zitadelUser = $client->fetchUserFromToken($accessToken);
+
+                $existingUser = $this->entityManager->getRepository(User::class)->findOneBy(
+                    ['oauthZitadelId' => $zitadelUser->getId()]
+                );
+
+                if ($existingUser) {
+                    return $existingUser;
+                }
+
+                $user = $this->userRepository->findOneBy(['email' => $zitadelUser->getEmail()]);
+
+                if ($user) {
+                    $user->oauthZitadelId = $zitadelUser->getId();
+
+                    $this->entityManager->flush();
+
+                    return $user;
+                }
+
+                $email = $zitadelUser->toArray()['preferred_username'];
+                $username = $slugger->slug(substr($email, 0, strrpos($email, '@')));
+
+                if ($this->userRepository->count(['username' => $username]) > 0) {
+                    $username .= rand(1, 999);
+                }
+
+                $dto = (new UserDto())->create(
+                    $username,
+                    $zitadelUser->getEmail()
+                );
+
+                $avatar = $this->getAvatar($zitadelUser->getPictureUrl());
+
+                if ($avatar) {
+                    $dto->avatar = $this->imageFactory->createDto($avatar);
+                }
+
+                $dto->plainPassword = bin2hex(random_bytes(20));
+                $dto->ip = $this->ipResolver->resolve();
+
+                $user = $this->userManager->create($dto, false);
+                $user->oauthZitadelId = $zitadelUser->getId();
+                $user->avatar = $this->getAvatar($zitadelUser->getPictureUrl());
+                $user->isVerified = true;
+
+                $this->entityManager->persist($user);
+                $this->entityManager->flush();
+
+                return $user;
+            }),
+            [
+                $rememberBadge,
+            ]
+        );
+    }
+
+    private function getAvatar(?string $pictureUrl): ?Image
+    {
+        if (!$pictureUrl) {
+            return null;
+        }
+
+        try {
+            $tempFile = $this->imageManager->download($pictureUrl);
+        } catch (\Exception $e) {
+            $tempFile = null;
+        }
+
+        if ($tempFile) {
+            $image = $this->imageRepository->findOrCreateFromPath($tempFile);
+            if ($image) {
+                $this->entityManager->persist($image);
+                $this->entityManager->flush();
+            }
+        }
+
+        return $image ?? null;
+    }
```

Comment:
Could you refactor this, so that it is not duplicated in 3 authenticators (Zitadel, Google and Facebook)

Reply:
I don't mind taking a shot at that when I go in to add the disable sso registration stuff as a follow up